### PR TITLE
Implement `equivalence` 

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Reference: https://github.com/lfortran/lfortran/pull/2558

### CODE REVIEW

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 1]
```diff
@@ -1486,6 +1487,13 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
 
         Vec<ASR::stmt_t*> body;
         body.reserve(al, x.n_body);
+        if(data_structure.size()>0){
```
Comment (Shaikh-Ubaid) [inline src/lfortran/semantics/ast_body_visitor.cpp:1490]:
```suggestion
        if (data_structure.size()>0){
```

Comment (certik) [reply src/lfortran/semantics/ast_body_visitor.cpp:1490]: Put a space on the right of `)` also.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -834,13 +834,14 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
-            std::map<std::string, std::vector<int>> &entry_function_arguments_mapping)
+            std::map<std::string, std::vector<int>> &entry_function_arguments_mapping,
+            std::vector<ASR::stmt_t*> &data_structure)
         : diag{diagnostics}, al{al}, compiler_options{compiler_options},
           current_scope{symbol_table}, implicit_mapping{implicit_mapping},
           common_variables_hash{common_variables_hash}, external_procedures_mapping{external_procedures_mapping},
           entry_functions{entry_functions},entry_function_arguments_mapping{entry_function_arguments_mapping},
           current_variable_type_{nullptr}, instantiate_types{instantiate_types},
-          instantiate_symbols{instantiate_symbols} {
+          instantiate_symbols{instantiate_symbols}, data_structure{data_structure}{
```
Comment (Shaikh-Ubaid) [inline src/lfortran/semantics/ast_common_visitor.h:844]:
```suggestion
          instantiate_symbols{instantiate_symbols}, data_structure{data_structure} {
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -796,7 +795,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
 
     std::map<std::string, ASR::ttype_t*> implicit_dictionary;
     std::map<uint64_t, std::map<std::string, ASR::ttype_t*>> &implicit_mapping;
-
+
```
Comment (Shaikh-Ubaid) [inline src/lfortran/semantics/ast_common_visitor.h:798]:
```suggestion

```

**Comment by HarshitaKalani** [comment]
This code still throws error with the following program:
```
program equivalence_05
    implicit none
    INTEGER IMACH(16), OUTPUT, SMALL(2)
    REAL RMACH
    EQUIVALENCE (IMACH(4),OUTPUT), (RMACH,SMALL(1))
    imach(4) = 6
    if(output /= 6) error stop
    small = 99
    rmach = 5.6_8
    if(small(1) /= 1085485875) error stop
    if(small(2) /= 99) error stop
    end
```

**Comment by HarshitaKalani** [comment]
With this, we're able to pass all the tests for `amos` with 0 patches!
https://github.com/certik/scipy/commits/merge_specfun_minpack_fitpack_amos_mach_02

**Review by certik — APPROVED** [review]
Great job! I think this is good enough to merge.

If there are any bugs, we can fix those later as we discover them.

Thanks!

**Pull request merged**